src/cache_manager.cc

```diff
@@ -0,0 +1,422 @@
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "cache_manager.h"
+#include <boost/functional/hash.hpp>
+#include "cache_entry.h"
+#include "filesystem.h"
+#include "server_message.h"
+#include "shared_library.h"
+#include "triton/common/logging.h"
+
+namespace triton { namespace core {
+
+std::string
+TritonCacheLibraryName()
+{
+#ifdef _WIN32
+  return std::string("tritoncache.dll");
+#else
+  return std::string("libtritoncache.so");
+#endif
+}
+
+//
+// TritonCache
+//
+Status
+TritonCache::Create(
+    const std::string& name, const std::string& libpath,
+    const std::string& cache_config, std::shared_ptr<TritonCache>* cache)
+{
+  LOG_INFO << "Creating TritonCache with name: '" << name << "', libpath: '"
+           << libpath << "', cache_config: '" << cache_config << "'";
+
+  auto lcache = std::shared_ptr<TritonCache>(
+      new TritonCache(name, libpath, cache_config));
+
+  RETURN_IF_ERROR(lcache->LoadCacheLibrary());
+  RETURN_IF_ERROR(lcache->InitializeCacheImpl());
+
+  *cache = std::move(lcache);
+  return Status::Success;
+}
+
+TritonCache::TritonCache(
+    const std::string& name, const std::string& libpath,
+    const std::string& cache_config)
+    : name_(name), libpath_(libpath), cache_config_(cache_config)
+{
+  ClearHandles();
+}
+
+TritonCache::~TritonCache()
+{
+  LOG_VERBOSE(1) << "unloading cache '" << name_ << "'";
+  if (fini_fn_) {
+    if (cache_impl_) {
+      LOG_VERBOSE(1) << "Calling TRITONCACHE_CacheFinalize from: '" << libpath_
+                     << "'";
+      LOG_TRITONSERVER_ERROR(fini_fn_(cache_impl_), "failed finalizing cache");
+    } else {
+      LOG_ERROR << "cache implementation handle is nullptr";
+    }
+  } else {
+    LOG_ERROR << "cache finalize function is nullptr";
+  }
+
+  ClearHandles();
+}
+
+void
+TritonCache::ClearHandles()
+{
+  dlhandle_ = nullptr;
```

Comment:
Need to close the library handle via SharedLibrary API

Reply:
done